### PR TITLE
applet.interface.uart: convert to nmigen

### PR DESCRIPTION
**Pull request by attie**
Addresses `interface.uart` item from #150.

### CODE REVIEW

**Comment by attie** [comment]
I'm currently unable to replicate the failures on my computer... will investigate later.

```
venv:$ python -W ignore::DeprecationWarning test.py -k 'glasgow.gateware.uart*'
.............
----------------------------------------------------------------------
Ran 13 tests in 0.823s

OK
```

**Comment by attie** [comment]
I'm aware of the failed test, but can't replicate it here... When running _all tests_, I see the following (same error, different test).

```
venv:$ python -W ignore::DeprecationWarning test.py
[...]
======================================================================
ERROR: test_build (glasgow.applet.memory.prom.MemoryPROMAppletTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/attie/proj_local/glasgow/glasgow/software/glasgow/applet/memory/prom/__init__.py", line 872, in test_build
    self.assertBuilds()
  File "/home/attie/proj_local/glasgow/glasgow/software/glasgow/applet/__init__.py", line 232, in assertBuilds
    target.build_plan().execute()
  File "/home/attie/proj_local/glasgow/glasgow/software/glasgow/target/hardware.py", line 118, in execute
    products  = self.lower.execute_local(build_dir)
  File "/home/attie/proj_local/glasgow/nmigen/nmigen/build/run.py", line 98, in execute_local
    subprocess.check_call(["sh", "{}.sh".format(self.script)])
  File "/home/attie/.bin/python3.8.2/lib/python3.8/subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['sh', 'build_top.sh']' returned non-zero exit status 1.

----------------------------------------------------------------------
Ran 238 tests in 168.288s

FAILED (errors=1, skipped=7)
```

When running the test specifically, I see:

```
venv:$ python -W ignore::DeprecationWarning test.py -vk glasgow.applet.interface.uart.UARTAppletTestCase
test_build (glasgow.applet.interface.uart.UARTAppletTestCase) ... Warning: Wire top.$flatten\i2c_target.\bus.$verilog_initial_trigger has an unprocessed 'init' attribute.
Warning: Wire top.$verilog_initial_trigger has an unprocessed 'init' attribute.
ok
test_loopback (glasgow.applet.interface.uart.UARTAppletTestCase) ... ok

----------------------------------------------------------------------
Ran 2 tests in 5.379s

OK
```

I have also seen errors such as below, when my code is bad. I was considering raising an issue, though not sure if it's nMigen or Glasgow.

```
venv:$ glasgow test uart
I: g.cli: testing applet 'uart'
test_build (glasgow.applet.interface.uart.UARTAppletTestCase) ...
ERROR: Can't open input file `/tmp/yosys-abc-3pHqzP/output.aig' for reading: No such file or directory
ERROR
Traceback (most recent call last):
  File "/home/attie/proj_local/glasgow/glasgow/software/glasgow/applet/interface/uart/__init__.py", line 381, in test_build
    self.assertBuilds()
  File "/home/attie/proj_local/glasgow/glasgow/software/glasgow/applet/__init__.py", line 232, in assertBuilds
    target.build_plan().execute()
  File "/home/attie/proj_local/glasgow/glasgow/software/glasgow/target/hardware.py", line 118, in execute
    products  = self.lower.execute_local(build_dir)
  File "/home/attie/proj_local/glasgow/nmigen/nmigen/build/run.py", line 98, in execute_local
    subprocess.check_call(["sh", "{}.sh".format(self.script)])
  File "/home/attie/.bin/python3.8.2/lib/python3.8/subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['sh', 'build_top.sh']' returned non-zero exit status 1.
```

**Comment by whitequark** [comment]
It is not your responsibility to fix the build error. (As in, I'll look into it, but not necessarily soon.)

**Comment by attie** [comment]
Ok, thanks.

The gateware and applet work, I've had to tweak things for the `nmigen.compat` -> `nmigen`, but I _think_ it should be fine as it is now.

**Comment by whitequark** [comment]
Okay, so the error is:
```
The command has to terminate. Boxes are not in a topological order.
The following information may help debugging (numbers are 0-based):
Input 0 of BoxA 1 (1stCI = 474; 1stCO = 1) has TFI with CI 509,
which corresponds to output 1 of BoxB 34 (1stCI = 508; 1stCO = 38).
In a correct topological order, BoxB should precede BoxA.
Error: Abc_CommandAbc9If(): Mapping of GIA has failed.
** cmd error: aborting 'source /tmp/yosys-abc-000000/abc.script'
ERROR: Can't open input file `/tmp/yosys-abc-000000/output.aig' for reading: No such file or directory
E.ABC command line: "source /tmp/yosys-abc-000000/abc.script".
```

Is your Yosys a very recent build from git? If not, that's probably why you don't see it. Try upgrading it, or use the CI configuration file to use YoWASP on your machine.

**Comment by whitequark** [comment]
@attie Can you rebase? I think post #251 the failure might be fixed.

File: software/glasgow/gateware/uart.py [hunk 1]
```diff
@@ -96,9 +101,12 @@ class UART(Module):
     """
     def __init__(self, pads, bit_cyc, data_bits=8, parity="none", max_bit_cyc=None,
                  invert_rx=False, invert_tx=False):
-        if max_bit_cyc is None:
-            max_bit_cyc = bit_cyc
-        self.bit_cyc = Signal(reset=bit_cyc, max=max_bit_cyc + 1)
+        self.bit_cyc = bit_cyc
+        self.data_bits = data_bits
+        self.parity = parity
+        self.max_bit_cyc = max_bit_cyc or bit_cyc
```
Comment (whitequark) [inline software/glasgow/gateware/uart.py:107]: This is not equivalent in all cases (`max_bit_cyc=0` would have different behavior). Although it makes little difference here, I strongly prefer explicit `is None`.

File: software/glasgow/gateware/uart.py [hunk 2]
```diff
@@ -131,170 +142,153 @@ def calc_parity(sig, kind):
...
         self.tx_o = self.tx_t.o
 
         self.bit_cyc = 4
-        self.submodules.dut = UART(pads=self, bit_cyc=self.bit_cyc)
+
+        self.dut = UART(pads=self, bit_cyc=self.bit_cyc)
+
+    def elaborate(self, platform):
+        m = Module()
+        m.submodules += self.dut
```
Comment (whitequark) [inline software/glasgow/gateware/uart.py:290]: I think you can just `return self.dut`.

**Review by whitequark — CHANGES_REQUESTED** [review]
See inline comments.

**Review by whitequark — DISMISSED** [review]
Code LGTM, not sure why tests are failing...

**Comment by attie** [comment]
My fault, I added a missing `.oe.eq(1)` (the UART Tx was non-functional), and did realise that I needed to update `Pin(width=1, dir='oe')` too.

**Review by whitequark — APPROVED** [review]

**Comment by whitequark** [comment]
Thanks!

**Pull request merged**